db/index.go

```diff
@@ -0,0 +1,495 @@
+	collection  client.Collection
+	desc        client.IndexDescription
+	convertFunc func(any) ([]byte, error)
+}
+
+var _ CollectionIndex = (*collectionSimpleIndex)(nil)
+
+func (i *collectionSimpleIndex) getDocKey(doc *client.Document) (core.IndexDataStoreKey, error) {
+	indexedFieldName := i.desc.Fields[0].Name
+	fieldVal, err := doc.Get(indexedFieldName)
+	isNil := false
+	if err != nil {
+		isNil = errors.Is(err, client.ErrFieldNotExist)
+		if !isNil {
+			return core.IndexDataStoreKey{}, nil // @todo: test
+		}
+	}
+
+	storeValue := ""
+	if isNil {
+		storeValue = indexFieldNilValue
+	} else {
+		data, err := i.convertFunc(fieldVal)
+		if err != nil {
+			return core.IndexDataStoreKey{}, NewErrCanNotIndexInvalidFieldValue(err)
+		}
+		storeValue = indexFieldValuePrefix + string(data)
+	}
+	indexDataStoreKey := core.IndexDataStoreKey{}
+	indexDataStoreKey.CollectionID = strconv.Itoa(int(i.collection.ID()))
+	indexDataStoreKey.IndexID = strconv.Itoa(int(i.desc.ID))
+	indexDataStoreKey.FieldValues = []string{storeValue, indexFieldValuePrefix + doc.Key().String()}
+	return indexDataStoreKey, nil
+}
+
+func (i *collectionSimpleIndex) Save(
+	ctx context.Context,
+	txn datastore.Txn,
+	doc *client.Document,
+) error {
+	key, err := i.getDocKey(doc)
+	if err != nil {
+		return err
+	}
+	err = txn.Datastore().Put(ctx, key.ToDS(), []byte{})
+	if err != nil {
+		return NewErrFailedToStoreIndexedField(key.IndexID, err)
+	}
+	return nil
+}
+
+func (i *collectionSimpleIndex) Update(
+	ctx context.Context,
+	txn datastore.Txn,
+	oldDoc *client.Document,
+	newDoc *client.Document,
+) error {
+	key, err := i.getDocKey(oldDoc)
+	if err != nil {
+		return err
+	}
+	err = txn.Datastore().Delete(ctx, key.ToDS())
+	if err != nil {
+		return err
+	}
+	return i.Save(ctx, txn, newDoc)
+}
+
+func (i *collectionSimpleIndex) RemoveAll(ctx context.Context, txn datastore.Txn) error {
+	prefixKey := core.IndexDataStoreKey{}
+	prefixKey.CollectionID = strconv.Itoa(int(i.collection.ID()))
+	prefixKey.IndexID = strconv.Itoa(int(i.desc.ID))
+	q, err := txn.Datastore().Query(ctx, query.Query{
+		Prefix: prefixKey.ToString(),
+	})
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := q.Close(); err != nil {
+			log.ErrorE(ctx, "Failed to close collection query", err)
+		}
+	}()
+
+	for res := range q.Next() {
+		if res.Error != nil {
+			return res.Error
+		}
+		err = txn.Datastore().Delete(ctx, ds.NewKey(res.Key))
+		if err != nil {
+			return NewCanNotDeleteIndexedField(err)
+		}
+	}
+
+	return nil
+}
+
+func (i *collectionSimpleIndex) Name() string {
+	return i.desc.Name
+}
+
+func (i *collectionSimpleIndex) Description() client.IndexDescription {
+	return i.desc
+}
+
+func validateIndexDescription(desc client.IndexDescription) error {
+	if desc.ID != 0 {
+		return NewErrNonZeroIndexIDProvided(desc.ID)
+	}
+	if len(desc.Fields) == 0 {
+		return ErrIndexMissingFields
+	}
+	if len(desc.Fields) == 1 && desc.Fields[0].Direction == client.Descending {
```

Comment:
question: We can't have a descending order if we only have one field for the index?

Reply:
that's correct. Different direction are only for composite indexes